In [ ]:
import psycopg2

In [ ]:
%load_ext sql

In [ ]:
%sql postgresql://kxxnuwbc:KM6wZP7RDKwXNOJ_3s78sAoAnUDUxxb0@trumpet.db.elephantsql.com:5432/kxxnuwbc

In [ ]:
%%sql
select table_name
from information_schema.tables
where table_schema = 'public'

 * postgresql://kxxnuwbc:***@trumpet.db.elephantsql.com:5432/kxxnuwbc
6 rows affected.


[('dim_geography',),
 ('dim_customer',),
 ('dim_product',),
 ('fact_sales',),
 ('dim_territory',),
 ('pg_stat_statements',)]

## Question 1

Untuk mengetahui target customer BeeCycle.
Kamu melakukan query untuk melihat distribusi customer berdasarkan gender dan umur.



Dimana untuk category umur, kamu membagi umur customer kedalam  (Hint : CASE WHEN)
* umur customer <= 20 tahun maka **'Group <=20'**
* umur customer antara 21 dan 40 tahun maka **'Group 21 - 40'**
* umur customer antara 41 dan 60 tahun maka **'Group 41 - 60'**
* umur customer lebih dari 60 tahun maka **'Group >60'**

**Grouping age apa dan gender apa yg memiliki transaksi paling tinggi di BeeCycle?**

In [ ]:
%%sql
select p.gender,
  case
    when p.age <= 20 then 'Group <=20'
    when p.age >= 21 and p.age <= 40 then 'Group 21 - 40'
    when p.age >= 41 and p.age <= 60 then 'Group 41 - 60'
    when p.age > 60 then 'Group >60'
    end as age_group,
    sum(fs.quantity) as purchase_qty
from dim_customer as p
join fact_sales as fs
on p.customer_id = fs.customer_id
group by p.gender, age_group
order by 3 desc

 * postgresql://kxxnuwbc:***@trumpet.db.elephantsql.com:5432/kxxnuwbc
6 rows affected.


[('F', 'Group 21 - 40', 1551),
 ('F', 'Group 41 - 60', 1454),
 ('M', 'Group 21 - 40', 1401),
 ('M', 'Group 41 - 60', 1304),
 ('M', 'Group >60', 161),
 ('F', 'Group >60', 82)]

Base on purchase_qty, transaksi tertinggi dipegang oleh gender Female dengan age_grup 21-40 dengan purchase qty 1551

In [ ]:
%%sql
select a.gender, a.group_age, count(*) as purchase_frequency
from (
  select gender,
  case
    when age <= 20 then 'Group <=20'
    when age >=21 and age <=40 then 'Group 21-40'
    when age >=41 and age <=60 then 'Group 41-60'
    when age >60 then 'Group >60'
  end as group_age
from dim_customer
) as a
group by gender, group_age
order by 3 desc

 * postgresql://kxxnuwbc:***@trumpet.db.elephantsql.com:5432/kxxnuwbc
6 rows affected.


[('F', 'Group 21-40', 361),
 ('M', 'Group 21-40', 349),
 ('F', 'Group 41-60', 313),
 ('M', 'Group 41-60', 289),
 ('M', 'Group >60', 27),
 ('F', 'Group >60', 21)]

Base on purchase_frequency, transaksi tertinggi dipegang oleh gender Female dengan age_grup 21-40 dengan total transaksi 361

In [ ]:
%%sql
select p.gender,
  case
    when p.age <= 20 then 'Group <=20'
    when p.age >= 21 and p.age <= 40 then 'Group 21 - 40'
    when p.age >= 41 and p.age <= 60 then 'Group 41 - 60'
    when p.age > 60 then 'Group >60'
    end as age_group,
    sum(fs.totalprice_rupiah) as totalprice_purchase
from dim_customer as p
join fact_sales as fs
on p.customer_id = fs.customer_id
group by p.gender, age_group
order by 3 desc

 * postgresql://kxxnuwbc:***@trumpet.db.elephantsql.com:5432/kxxnuwbc
6 rows affected.


[('F', 'Group 21 - 40', 20994324000.0),
 ('M', 'Group 21 - 40', 19722105000.0),
 ('F', 'Group 41 - 60', 18316800000.0),
 ('M', 'Group 41 - 60', 15888866000.0),
 ('M', 'Group >60', 1256616800.0),
 ('F', 'Group >60', 999984060.0)]

Base on totalprice, transaksi tertinggi dipegang oleh gender Female dengan age_grup 21-40 dengan total transaksi hingga 20.994.324.000

## Question 2

Karena bulan Oktober depan akan ada campaign.
Kamu punya ide untuk menghilight warna product tertentu yg menjadi kegemaran customer. Kamu selanjutnya melakukan query untuk menjawab pertanyaan berikut :

Sebenarnya saya agak bingung memahami instruksi ini, karena ada dua makna:<br>
A. Mengambil product tertentu kemudian mencari warna kegemaran customer dari product tersebut, atau<br>
B. Mengambil semua product dan mengurutkanya base on color <br>
Maka dari itu saya membuat 2 query


Query A

In [ ]:
%%sql
select *
from dim_product
limit 3

 * postgresql://kxxnuwbc:***@trumpet.db.elephantsql.com:5432/kxxnuwbc
3 rows affected.


[(210, 'HL Road Frame - Black, 58', 'HL Road Frame', '54-58 CM', 'Black', 11000.0, 11000.0, 'Road Frames', 'Components', 20),
 (211, 'HL Road Frame - Red, 58', 'HL Road Frame', '54-58 CM', 'Red', 11000.0, 11000.0, 'Road Frames', 'Components', 20),
 (480, 'Patch Kit/8 Patches', 'Patch kit', 'NA', 'NA', 11991.0, 32060.0, 'Tires and Tubes', 'Accessories', 20)]

Kebanyakan kolom 'product_name' hanya memiliki 1 warna karena nama product sekaligus mencantumkan warnanya, contohnya 'HL Road Frame - Red, 58' yang hanya memiliki warna merah. <br>
Maka saya mencobanya dengan mengganti dengan sub category dimana variasi warnanya akan semakin banyak

In [ ]:
%%sql
select a.color, sum(b.quantity) as total_order_qty
from dim_product as a
join fact_sales as b
on a.product_id = b.product_id
where a.sub_category='Helmets'
group by a.color
order by total_order_qty desc

 * postgresql://kxxnuwbc:***@trumpet.db.elephantsql.com:5432/kxxnuwbc
3 rows affected.


[('Blue', 221), ('Red', 201), ('Black', 175)]

Penjualan product dengan kategory "Helmet" warna Blue sangat digemari customer

Query B

In [ ]:
%%sql
select *
from fact_sales
limit 3

 * postgresql://kxxnuwbc:***@trumpet.db.elephantsql.com:5432/kxxnuwbc
3 rows affected.


[('SO43698-1', '2016-07-01', 346, 28389, 7, 'SO43698', 1, 1, 47599860.0, 47599860.0, 26770162.0, 1189997.0, 20829698.0),
 ('SO43704-1', '2016-07-02', 351, 11005, 9, 'SO43704', 1, 1, 47249860.0, 47249860.0, 26573322.0, 1181247.0, 20676538.0),
 ('SO43705-1', '2016-07-02', 344, 11011, 9, 'SO43705', 1, 1, 47599860.0, 47599860.0, 26770162.0, 1189997.0, 20829698.0)]

In [ ]:
%%sql
select a.color, sum(b.quantity) as total_order_qty
from dim_product as a
join fact_sales as b
on a.product_id = b.product_id
group by a.color
order by total_order_qty desc

 * postgresql://kxxnuwbc:***@trumpet.db.elephantsql.com:5432/kxxnuwbc
8 rows affected.


[('NA', 2355),
 ('Black', 1051),
 ('Silver', 648),
 ('Red', 646),
 ('Blue', 502),
 ('Yellow', 419),
 ('Multi', 288),
 ('White', 45)]

Dari data dapat dilihat bahwa product favorit customer peringkat pertama tidak tercantum warnanya, sedangkan product yang diketahui warnanya dan menempati peringkat teratas adalah "Black". maka dapat disimpulkan warna product favorit customer adalah "Black"

## Question 3

Karena campaign bulan Oktober tsb hanya terbatas untuk tiap2 toko cabang.
kamu punya ide, bagaimana hanya mengambil TOP 10 product_id dari masing2 toko cabang. sehingga kamu melakukan query untuk menjawab pertanyaan berikut :

In [ ]:
%%sql
with ranked_products as (
select
  a.product_id,
  b.region,
  sum(c.sales_order_line_number) as total_sales,
  row_number() over (partition by b.region order by sum(c.sales_order_line_number) desc) as rank
from
  fact_sales c
  join dim_product a on c.product_id = a.product_id
  join dim_territory b on c.territory_id = b.territory_id
group by
  a.product_id, b.region
)

select
  region,
  product_id,
  total_sales
from ranked_products
where rank <= 10
order by 1,3 desc

 * postgresql://kxxnuwbc:***@trumpet.db.elephantsql.com:5432/kxxnuwbc
75 rows affected.


[('Australia', 477, 322),
 ('Australia', 214, 186),
 ('Australia', 478, 177),
 ('Australia', 480, 173),
 ('Australia', 222, 168),
 ('Australia', 479, 150),
 ('Australia', 485, 145),
 ('Australia', 225, 142),
 ('Australia', 217, 141),
 ('Australia', 487, 119),
 ('Canada', 480, 183),
 ('Canada', 529, 113),
 ('Canada', 528, 84),
 ('Canada', 217, 81),
 ('Canada', 222, 70),
 ('Canada', 477, 69),
 ('Canada', 214, 62),
 ('Canada', 484, 61),
 ('Canada', 225, 58),
 ('Canada', 530, 53),
 ('France', 477, 153),
 ('France', 222, 107),
 ('France', 479, 89),
 ('France', 214, 80),
 ('France', 478, 71),
 ('France', 480, 68),
 ('France', 225, 64),
 ('France', 217, 60),
 ('France', 530, 37),
 ('France', 485, 33),
 ('Germany', 477, 144),
 ('Germany', 225, 110),
 ('Germany', 478, 98),
 ('Germany', 217, 81),
 ('Germany', 222, 70),
 ('Germany', 214, 63),
 ('Germany', 479, 55),
 ('Germany', 485, 41),
 ('Germany', 480, 31),
 ('Germany', 541, 30),
 ('Northwest', 480, 65),
 ('Northwest', 537, 64),
 ('Northwest', 528, 52),
 ('Northwest', 477, 41),
 ('Northwest', 485, 38),
 ('Northwest', 222, 32),
 ('Northwest', 225, 28),
 ('Northwest', 310, 26),
 ('Northwest', 478, 23),
 ('Northwest', 217, 22),
 ('Southeast', 477, 4),
 ('Southeast', 473, 3),
 ('Southeast', 478, 2),
 ('Southeast', 480, 1),
 ('Southeast', 353, 1),
 ('Southwest', 480, 98),
 ('Southwest', 537, 84),
 ('Southwest', 485, 82),
 ('Southwest', 528, 66),
 ('Southwest', 477, 60),
 ('Southwest', 478, 55),
 ('Southwest', 217, 44),
 ('Southwest', 225, 39),
 ('Southwest', 312, 37),
 ('Southwest', 487, 34),
 ('United Kingdom', 477, 171),
 ('United Kingdom', 225, 135),
 ('United Kingdom', 214, 124),
 ('United Kingdom', 222, 123),
 ('United Kingdom', 478, 87),
 ('United Kingdom', 479, 86),
 ('United Kingdom', 217, 60),
 ('United Kingdom', 480, 54),
 ('United Kingdom', 530, 44),
 ('United Kingdom', 541, 42)]

## Question 4

Karena sudah lulus probation nih.. kamu diminta untuk membuat sendiri analysis untuk si Bos. Beberapa step yg perlu dilakukan adalah :
1. Membuat objective analysis (contoh : Karena next campaign mau dibuat maksimal , kita akan menghilight top2 product)
2. Membuat business question (contoh : Apa saja product yg paling populer dari segi jumlah transaksi?)
3. Build the Query untuk menjawab pertanyaan pada poin no 2
4. Tulis Actionable knowledge atau next action setelah mendapat jawaban dari poin 3 (contoh : product abc yg paling tinggi sehingga kita akan fokus untuk promosi product tersebut)

1. Karena next campaign dibuat maksimal, kita akan menghilight top 3 product berdasarkan  klasifikasi umur

2. Apa saja product yang paling populer berdasarkan frekuensi transaksi yang dilakukan dari range umur tertentu?

3. Query

In [ ]:
%%sql
with Top3 as
  (select product_name, age_group, freq_purchase,
    dense_rank() over (partition by age_group order by freq_purchase desc) as age_group_rank
  from
    (select
      c.product_name,
      case
        when age <= 20 THEN 'Group <=20'
        when age >=21 and age <=25 then 'Group 21-25'
        when age >=26 and age <=30 then 'Group 26-30'
        when age >=31 and age <=35 then 'Group 31-35'
        when age >=36 and age <=40 then 'Group 36-40'
        when age >=41 and age <=45 then 'Group 41-45'
        when age >=46 and age <=50 then 'Group 46-50'
        when age >=51 and age <=55 then 'Group 51-55'
        when age >=56 and age <=60 then 'Group 56-60'
        when age > 60 then 'Group >60'
      end as age_group,
      count(fs.quantity) as freq_purchase
    from dim_customer as p
    join fact_sales as fs on p.customer_id = fs.customer_id
    join dim_product as c on fs.product_id = c.product_id
    group by c.product_name, age_group)
    as AgeGroupData)

select age_group, product_name, freq_purchase
from Top3
where age_group_rank <= 3
order by 3 desc

 * postgresql://kxxnuwbc:***@trumpet.db.elephantsql.com:5432/kxxnuwbc
33 rows affected.


[('Group 31-35', 'Water Bottle - 30 oz.', 70),
 ('Group 46-50', 'Water Bottle - 30 oz.', 64),
 ('Group 41-45', 'Water Bottle - 30 oz.', 64),
 ('Group 36-40', 'Water Bottle - 30 oz.', 46),
 ('Group 31-35', 'Mountain Bottle Cage', 44),
 ('Group 26-30', 'Water Bottle - 30 oz.', 43),
 ('Group 41-45', 'Patch Kit/8 Patches', 43),
 ('Group 36-40', 'Patch Kit/8 Patches', 43),
 ('Group 41-45', 'Mountain Bottle Cage', 42),
 ('Group 31-35', 'Patch Kit/8 Patches', 40),
 ('Group 46-50', 'Mountain Bottle Cage', 39),
 ('Group 46-50', 'Sport-100 Helmet, Blue', 37),
 ('Group 36-40', 'HL Mountain Tire', 35),
 ('Group 51-55', 'Water Bottle - 30 oz.', 34),
 ('Group 26-30', 'Mountain Bottle Cage', 34),
 ('Group 26-30', 'Mountain-200 Silver, 38', 32),
 ('Group 51-55', 'Sport-100 Helmet, Blue', 27),
 ('Group 51-55', 'Patch Kit/8 Patches', 27),
 ('Group 56-60', 'Patch Kit/8 Patches', 26),
 ('Group 56-60', 'Road Bottle Cage', 20),
 ('Group >60', 'Patch Kit/8 Patches', 20),
 ('Group 51-55', 'Mountain Bottle Cage', 20),
 ('Group 56-60', 'Mountain Tire Tube', 19),
 ('Group 56-60', 'HL Mountain Tire', 19),
 ('Group 56-60', 'Road Tire Tube', 19),
 ('Group >60', 'Mountain Tire Tube', 18),
 ('Group >60', 'Water Bottle - 30 oz.', 13),
 ('Group >60', 'HL Mountain Tire', 13),
 ('Group 21-25', 'Patch Kit/8 Patches', 12),
 ('Group 21-25', 'Fender Set - Mountain', 12),
 ('Group 21-25', 'Water Bottle - 30 oz.', 11),
 ('Group 21-25', 'Mountain Tire Tube', 11),
 ('Group 21-25', 'HL Mountain Tire', 10)]

4. "Water Bottle - 30 oz."	sangat diminati oleh customer dengan range umur 31 sampai 50, dan menempati 5 product terlaris <br>
Maka dari itu penjualan dengan product serupa tetapi dengan varisai yang lebih beragam akan dilakukan, serta promosi akan lebih difokuskan untuk product terseut